File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
     int ans;
        int  solve(TreeNode root){
              if(root==null){
                   return 0;
              }
                int left= solve(root.left);
                int right= solve(root.right);

                ans= Math.max(left + right, ans );

                return 1+ Math.max(left, right );
        }
    public int diameterOfBinaryTree(TreeNode root) {
          ans=0;

        solve(root);


       return ans;
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {  

public:     
   bool solve(TreeNode *p, TreeNode* q){
                if(p==NULL && q==NULL){
                     return true;
                }

                if(p==NULL || q==NULL){
                    return false;
                }
                  
                return solve(p->left, q->left) && solve(p->right, q->right) &&(p->val== q->val);

        }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
           bool  flag= false;
            queue<TreeNode*>q;
                if(root==NULL   && subRoot==NULL){
                    return true;
                }
                if(root==NULL || subRoot==NULL){
                    return false;
                }
            q.push(root);

            while(!q.empty()){
                  TreeNode *curr= q.front();
                  q.pop();

                  if(curr->val==subRoot->val){
                      flag= flag || solve(curr, subRoot);
                  }

                  if(curr->left){
                    q.push(curr->left);
                  }
                  if(curr->right){
                    q.push(curr->right);
                  }
            }

            return flag;
    }
};
